custom/benin/management/commands/migrate_users_and_their_cases_to_new_rc_level.py

```diff
@@ -0,0 +1,189 @@
+from corehq.util.log import with_progress_bar
+
+LOCATION_TYPE_VILLAGE = "Village"
+LOCATION_TYPE_RC = "RC"
+
+
+logfile = f"migrate_users_and_their_cases_to_new_rc_level_{datetime.datetime.utcnow()}"
+
+
+class Command(BaseCommand):
+    help = 'Migrate benin project\'s users and their cases to new rc level locations'
+
+    def add_arguments(self, parser):
+        super(Command, self).add_arguments(parser)
+
+        parser.add_argument('domain')
+        parser.add_argument(
+            '--village_id'
+        )
+        parser.add_argument(
+            '--dry_run',
+            action='store_true',
+            default=False,
+            help="A dry run to only share the updates that would happen",
+        )
+
+    def handle(self, domain, **options):
+        """
+        Steps:
+        1. Fetch all villages (location type: Village).
+        2. For each village:
+            1. Fetch all users assigned to the village with usertype 'rc'
+            2. For each user
+                1. find the corresponding RC under the village with name same as user's user data in rc_number.
+                   Log error if no matching RC, and move to next user
+                2. if RC present
+                    1. Find all OPEN cases (case_type: menage, membre)
+                        1. owned by village
+                        2. opened_by the user (Use CaseES.opened_by)
+                    2. Find all OPEN cases (case_type: seance_educative, fiche_pointage)
+                        1. opened_by the user (Use CaseES.opened_by)
+                        2. Why are we updating it though? They are already owned by users. They aren't many though
+                    3. Update all cases
+                        1. Update owner to be the corresponding RC location
+                    4. Update users location to corresponding RC location only after cases to enable
+                       retry on this update in case of any intermittent failures
+        """
+        dry_run = options['dry_run']
+        village_id = options['village_id']
+
+        if village_id:
+            villages = SQLLocation.active_objects.get_locations([village_id])
+        else:
+            villages = _find_locations(domain=domain, location_type_code=LOCATION_TYPE_VILLAGE)
+        for village in villages:
+            log(f"Starting updates for village {village.name}")
+            users = _find_rc_users_at_location(domain, village)
+            for user in users:
+                user_rc_number = user.user_data.get('rc_number')
+                if user_rc_number:
+                    try:
+                        new_user_rc_location = _find_child_location_with_name(
+                            parent_location=village,
+                            location_name=user_rc_number
+                        )
+                    except MultipleMatchingLocationsFound:
+                        log(f"Multiple matching locations found for user {user.username}:{user.user_id} "
```

Comment:
Should we log errors to a separate file? I'm just thinking in terms of visibility so that it's easier to see if anything went wrong.

Reply:
Yup, https://github.com/dimagi/commcare-hq/pull/34588/commits/5bf510b934755c72ae7e5fdb31226622e05ee98e